File: src/main/java/com/learning/leetcode/recursive/medium/combinations/CombinationsSum2.java
package com.learning.leetcode.recursive.medium.combinations;

import java.util.*;

public class CombinationsSum2 {

    public static void main(String[] args) {
        final List<List<Integer>> result = new ArrayList<>();
        final int[] candidates = {10, 1, 2, 7, 6, 1, 5};
        final int target = 8;
        Arrays.sort(candidates);
        combinations(candidates, target, 0, 0, new ArrayList<>(), result);
        System.out.println("No of combinations for " + Arrays.toString(candidates) + " and target " + target + " are " + result.size());
        result.forEach(System.out::println);
    }

    static void combinations(final int[] nums, final int target, int index, int sumTillNow, final List<Integer> subset, final List<List<Integer>> result) {

        if (sumTillNow == target) {
            result.add(new ArrayList<>(subset));
            return;
        }
        if (sumTillNow > target) {
            return;
        }
        if (index == nums.length) {
            return;
        }

        subset.add(nums[index]);
        sumTillNow += nums[index];
        combinations(nums, target, index + 1, sumTillNow, subset, result);
        sumTillNow -= nums[index];
        subset.remove(subset.size() - 1);
        while (index + 1 < nums.length && nums[index] == nums[index + 1])
            index++;
        combinations(nums, target, index + 1, sumTillNow, subset, result);
    }
}

File: src/main/java/com/learning/leetcode/recursive/medium/combinations/CombinationsSum2.py
from typing import List


def combinationSum(arr: List[int], index: int, sumSoFar: int, target: int, path: List[int], result: List[List[int]]):
    if sumSoFar == target:
        result.append(path[:])
        return
    if sumSoFar > target or index == len(arr):
        return

    path.append(arr[index])
    sumSoFar += arr[index]
    combinationSum(arr, index + 1, sumSoFar, target, path, result)
    sumSoFar -= arr[index]
    path.pop()
    while index + 1 < len(arr) and arr[index] == arr[index + 1]:
        index += 1
    combinationSum(arr, index + 1, sumSoFar, target, path, result)


def combinations(nums, target, index, sum_till_now, subset, result):
    if sum_till_now == target:
        result.append(subset.copy())
        return
    if not sum_till_now > target or not index == len(nums):
        subset.append(nums[index])
        sum_till_now += nums[index]
        combinations(nums, target, index + 1, sum_till_now, subset, result)
        sum_till_now -= nums[index]
        subset.pop()

        while index + 1 < len(nums) and nums[index] == nums[index + 1]:
            index += 1

        combinations(nums, target, index + 1, sum_till_now, subset, result)


def main():
    nums = [10, 1, 2, 7, 6, 1, 5]
    nums.sort()
    target = 8
    path, result = [], []
    combinations(nums, target, 0, 0, path, result)
    #combinationSum(nums, 0, 0, target, path, result)
    print(result)


main()
